src/auth_context.rs

```diff
@@ -0,0 +1,172 @@
+//! API for authenticating peer
+//! Based on https://grpc.github.io/grpc/core/md_doc_server_side_auth.html
+
+use std::ffi::CStr;
+use std::marker::PhantomData;
+
+use crate::grpc_sys::{
+    self, GrpcAuthContext, GrpcAuthProperty, GrpcAuthPropertyIterator, GrpcCall,
+};
+
+/// To perform server-side authentication, gRPC exposes the authentication context
+/// for each call. The context exposes important authentication-related information
+/// about the RPC such as the type of security/authentication type being used and
+/// the peer identity.
+///
+/// The authentication context is structured as a multi-map of key-value pairs -
+/// the auth properties. In addition to that, for authenticated RPCs, the set of
+/// properties corresponding to a selected key will represent the verified identity
+/// of the caller - the peer identity.
+///
+/// The contents of the auth properties are populated by an auth interceptor within
+/// gRPC Core. The interceptor also chooses which property key will act as the peer
+/// identity (e.g. for client certificate authentication this property will be
+/// `x509_common_name` or `x509_subject_alternative_name`).
+pub struct AuthContext<'a> {
```

Comment:
I've set this lifetime because I didn't know whether stuff would get deallocated when the call would get deallocated.

Looking at the grpc source code it would seem that this is in fact RCed:
https://github.com/grpc/grpc/blob/4fa4f8e74b2b82a8991ef5e714c3b7c822682eb4/src/core/lib/security/context/security_context.h#L134

Can you confirm that I can for sure remove this lifetime?

Reply:
Yes, it's a subclass of `RefCounted`.